File: diagramas/ChefApp/ChefApp/app/src/main/java/com/chefapp/fragments/BotonesFragment.java
package com.chefapp.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.chefapp.R;
import com.google.android.material.switchmaterial.SwitchMaterial;

public class BotonesFragment extends Fragment {

    private int currentRating = 4;
    private int currentDifficulty = 1; // 0=Fácil, 1=Medio, 2=Difícil
    private ImageView[] stars;
    private TextView btnFacil, btnMedio, btnDificil;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_botones, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        setupRating(view);
        setupUnits(view);
        setupDifficulty(view);
        setupPorciones(view);
        setupRestricciones(view);
    }

    private void setupRating(View view) {
        stars = new ImageView[]{
                view.findViewById(R.id.star1),
                view.findViewById(R.id.star2),
                view.findViewById(R.id.star3),
                view.findViewById(R.id.star4),
                view.findViewById(R.id.star5)
        };
        TextView tvLabel = view.findViewById(R.id.tv_rating_label);

        for (int i = 0; i < stars.length; i++) {
            final int rating = i + 1;
            stars[i].setOnClickListener(v -> {
                currentRating = rating;
                updateStars(tvLabel);
            });
        }
        updateStars(tvLabel);
    }

    private void updateStars(TextView tvLabel) {
        String[] labels = {"", "Regular", "Bueno", "Muy bueno", "Muy bueno", "Excelente"};
        for (int i = 0; i < stars.length; i++) {
            stars[i].setImageResource(i < currentRating ?
                    R.drawable.ic_star_filled : R.drawable.ic_star_outline);
        }
        if (currentRating >= 0 && currentRating <= 5) {
            tvLabel.setText(labels[currentRating]);
        }
    }

    private void setupUnits(View view) {
        SwitchMaterial switchUnits = view.findViewById(R.id.switch_units);
        switchUnits.setOnCheckedChangeListener((buttonView, isChecked) ->
                Toast.makeText(getContext(),
                        isChecked ? "Onzas / °F" : "Gramos / °C",
                        Toast.LENGTH_SHORT).show());
    }

    private void setupDifficulty(View view) {
        btnFacil = view.findViewById(R.id.btn_facil);
        btnMedio = view.findViewById(R.id.btn_medio);
        btnDificil = view.findViewById(R.id.btn_dificil);

        btnFacil.setOnClickListener(v -> selectDifficulty(0));
        btnMedio.setOnClickListener(v -> selectDifficulty(1));
        btnDificil.setOnClickListener(v -> selectDifficulty(2));

        selectDifficulty(1); // Default: Medio
    }

    private void selectDifficulty(int index) {
        currentDifficulty = index;
        btnFacil.setBackgroundResource(index == 0 ? R.drawable.bg_difficulty_selected : R.drawable.bg_difficulty_normal);
        btnMedio.setBackgroundResource(index == 1 ? R.drawable.bg_difficulty_selected : R.drawable.bg_difficulty_normal);
        btnDificil.setBackgroundResource(index == 2 ? R.drawable.bg_difficulty_selected : R.drawable.bg_difficulty_normal);

        btnFacil.setTextColor(getResources().getColor(index == 0 ? R.color.text_white : R.color.text_primary, null));
        btnMedio.setTextColor(getResources().getColor(index == 1 ? R.color.text_white : R.color.text_primary, null));
        btnDificil.setTextColor(getResources().getColor(index == 2 ? R.color.text_white : R.color.text_primary, null));
    }

    private void setupPorciones(View view) {
        SeekBar seekBar = view.findViewById(R.id.seekbar_porciones);
        TextView tvValue = view.findViewById(R.id.tv_porciones_value);
        TextView tvTiempo = view.findViewById(R.id.tv_tiempo_estimado);

        seekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                int porciones = progress + 1;
                tvValue.setText(String.valueOf(porciones));
                int baseTime = 35;
                int estimatedTime = baseTime + (porciones - 1) * 5;
                tvTiempo.setText("Tiempo estimado: " + estimatedTime + " min");
            }
            @Override public void onStartTrackingTouch(SeekBar seekBar) {}
            @Override public void onStopTrackingTouch(SeekBar seekBar) {}
        });

        view.findViewById(R.id.btn_calcular).setOnClickListener(v -> {
            int porciones = seekBar.getProgress() + 1;
            String diff = currentDifficulty == 0 ? "Fácil" : currentDifficulty == 1 ? "Medio" : "Difícil";
            Toast.makeText(getContext(),
                    "Calculando para " + porciones + " porciones (" + diff + ")",
                    Toast.LENGTH_SHORT).show();
        });
    }

    private void setupRestricciones(View view) {
        CheckBox cbGluten = view.findViewById(R.id.cb_sin_gluten);
        CheckBox cbVegano = view.findViewById(R.id.cb_vegano);
        CheckBox cbLactosa = view.findViewById(R.id.cb_sin_lactosa);
        CheckBox cbMariscos = view.findViewById(R.id.cb_sin_mariscos);

        View.OnClickListener listener = v -> {
            StringBuilder sb = new StringBuilder("Filtros: ");
            if (cbGluten.isChecked()) sb.append("Sin gluten ");
            if (cbVegano.isChecked()) sb.append("Vegano ");
            if (cbLactosa.isChecked()) sb.append("Sin lactosa ");
            if (cbMariscos.isChecked()) sb.append("Sin mariscos ");
            Toast.makeText(getContext(), sb.toString().trim(), Toast.LENGTH_SHORT).show();
        };

        cbGluten.setOnClickListener(listener);
        cbVegano.setOnClickListener(listener);
        cbLactosa.setOnClickListener(listener);
        cbMariscos.setOnClickListener(listener);
    }
}

File: app/src/main/java/com/example/chefapp/fragments/BotonesFragment.kt
package com.example.chefapp.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.SeekBar
import androidx.fragment.app.Fragment
import com.example.chefapp.databinding.FragmentBotonesBinding
import com.example.chefapp.models.*

class BotonesFragment : Fragment() {

    private var _binding: FragmentBotonesBinding? = null
    private val binding get() = _binding!!

    private var recetaActual: Receta = RecetaData.recetas.first()

    private var porciones = recetaActual.porciones
    private var unidadActual = UnidadMedida.GRAMOS

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentBotonesBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.ratingBar.setOnRatingBarChangeListener { _, rating, _ ->
            calificarReceta(rating.toInt())
        }

        binding.switchUnidades.setOnCheckedChangeListener { _, isChecked ->
            cambiarUnidades(isChecked)
        }

        binding.radioGroupDificultad.setOnCheckedChangeListener { _, checkedId ->
            seleccionarRestricciones()
            val dificultad = when (checkedId) {
                binding.radioFacil.id   -> "Fácil ✅"
                binding.radioMedio.id   -> "Medio ⚠️"
                binding.radioDificil.id -> "Difícil 🔥"
                else -> ""
            }
            binding.tvDificultadSeleccionada.text = dificultad
        }

        binding.seekBarPorciones.max = 9
        binding.seekBarPorciones.progress = porciones - 1
        binding.tvPorciones.text = "$porciones porciones"

        binding.seekBarPorciones.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
                porciones = progress + 1
                binding.tvPorciones.text = "$porciones porciones"
            }
            override fun onStartTrackingTouch(seekBar: SeekBar?) {}
            override fun onStopTrackingTouch(seekBar: SeekBar?) {}
        })

        binding.btnCalcular.setOnClickListener {
            calcularIngredientes()
        }

        binding.checkGluten.setOnCheckedChangeListener { _, _ -> seleccionarRestricciones() }
        binding.checkVegano.setOnCheckedChangeListener { _, _ -> seleccionarRestricciones() }
        binding.checkLactosa.setOnCheckedChangeListener { _, _ -> seleccionarRestricciones() }
        binding.checkMariscos.setOnCheckedChangeListener { _, _ -> seleccionarRestricciones() }
    }

    private fun calificarReceta(valor: Int) {
        val etiqueta = when (valor) {
            1 -> "Malo 😞"
            2 -> "Regular 😐"
            3 -> "Bueno 🙂"
            4 -> "Muy bueno 😊"
            5 -> "Excelente 🤩"
            else -> ""
        }
        binding.tvRatingLabel.text = etiqueta
        UsuarioActual.usuario.calificarReceta(recetaActual, valor)
    }

    private fun cambiarUnidades(usarOnzas: Boolean) {
        unidadActual = if (usarOnzas) UnidadMedida.ONZAS else UnidadMedida.GRAMOS
        binding.tvUnidadActiva.text = if (unidadActual == UnidadMedida.GRAMOS) "Gramos / °C" else "Onzas / °F"
    }

    private fun seleccionarRestricciones() {
        val usuario = UsuarioActual.usuario
        usuario.alergias.clear()
        if (binding.checkGluten.isChecked)   usuario.agregarAlergia(Alergia(1, "Gluten"))
        if (binding.checkVegano.isChecked)   usuario.agregarPreferencia(PreferenciaDietetica(1, TipoDieta.VEGANA))
        if (binding.checkLactosa.isChecked)  usuario.agregarAlergia(Alergia(2, "Lactosa"))
        if (binding.checkMariscos.isChecked) usuario.agregarAlergia(Alergia(3, "Mariscos"))
    }

    private fun calcularIngredientes() {
        val recetaAjustada = recetaActual.ajustarPorciones(porciones)

        val resultado = if (unidadActual == UnidadMedida.GRAMOS) {
            recetaAjustada.ingredientes.joinToString(" | ") { "${it.nombre}: ${it.cantidad.toInt()}g" }
        } else {
            recetaAjustada.ingredientes.joinToString(" | ") { ingrediente ->
                val enOnzas = ingrediente.cantidad * 0.035274f
                "${ingrediente.nombre}: ${String.format("%.1f", enOnzas)}oz"
            }
        }

        binding.tvResultado.text = resultado
        binding.tvResultado.visibility = View.VISIBLE
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}